book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,215 @@
+    laptop: bcde...
+  HTTP
+    workstation: cdef...
+```
+
+Person accounts will need a new `Attribute::ApplicationPassword` that stores a
+`ValueSetApplicationPassword`. Each value in the set is a new type to manage these
+secrets and their labeling and the references to the applications.
+
+```
+type ApplicationPasswords = BTreeMap<Uuid, (String, Password)>;
+                                     ^        ^         ^
+                                     |        |         +-> The hashed password
+                                     |        +-> Application password label
+                                     +-> Application password UUID
+
+type ApplicationUuid = Uuid;
+
+struct ValueSetApplicationPassword {
+  map: Map<ApplicationUuid, ApplicationPasswords>
+}
+```
+
+Each value in the set is queried by its UUID. This defines the value as
+`Value::ApplicationPassword(Uuid, ApplicationPassword)`. The
+`ApplicationPassword` type implements `PartialEq` so two application passwords
+are equal if their label is equal and they refer to the same application.
+
+(NOTE: We could consider extending the `CredentialType` to add
+`CredentialType::ApplicationPassword<Map<Uuid, ApplicationPassword>>`. Then we
+wouldn't need `Attribute::ApplicationPassword`, nor `Value::ApplicationPassword`,
+neither `ValueSetApplicationPassword`).
+
+The user must be able to delete credentials individually. The generated password
+is only displayed once when the user creates it and it is not possible to
+recover the clear-text form, only hashed form is stored. It is not allowed to
+store duplicated application passwords (same app refer and label).
+
+We do not need temporary locks or holds - users can delete and recreate as
+needed.
+
+### Reference integrity
+
+Since application passwords are related to applications, on delete of an
+application all entries that have a bound application password should be removed
+from user accounts.
+
+### Access controls
+
+The IDM administrators group can manage the applications and user's application
+passwords.
+
+Users can manage their own application passwords.
+
+### LDAP
+
+The bind DN regular expression needs to adjusted to detect and determine the
+bind dn if it is related to an application or not. The application bind dn
+regular expression will capture the user name, the label and the application
+name.
+
+The rootdse needs to be extended to include the applications as additional
+naming contexts.
+
+If the session is related to an application we should only accept application
+passwords in the bind. The user needs to be a member of the application. Binds
+to an application password must be limited by account validity and expiration.
+
+We need to add a cache of available applications for lookup.
+
+(NOTE: Caching and reload needs more explanation)
+
+### kanidm CLI
+
+The `kanidm` command line tool will be extended with the following commands to
+manage the applications:
+
+* kanidm application create <NAME>
```

Comment:
applications also need service-account api-tokens with them. This is because the application needs to be able to search ldap with elevated read permissions potentially. 

So I think before we define what the CLI will look like, we need to look at what we may need to configure.

Reply:
Ok